### PR TITLE
Ensure auth

### PR DESCRIPTION
**Pull request by jupe**
replace https://github.com/OpenTMI/opentmi/pull/16

Our initial test seeding method (dookie) was erasing all the index definitions when it dumped the collections (part of the seeding process).

We are now seeding the database for testing via shell scripts that use Mongo's own import and export functions which retain index information.

Also the all the routes now use authorization, and the tests had to be updated to use authorization.

### CODE REVIEW

File: app/controllers/builds.js [hunk 1]
```diff
@@ -25,43 +26,45 @@ var Controller = function(){
...
-    if( yes === true ){
-      var Template = {
-        name: 'Build-',
-        target: {
-          type: 'simulate',
-          hw: {
-            platform: ''
+  if( nconf.get('seeds') ){
+    //create dummy testcases when db is empty ->
+    defaultCtrl.isEmpty( function(yes){
```
Comment (jupe) [inline app/controllers/builds.js]: remove whole dummy data section because its not in use

File: app/controllers/campaigns.js [hunk 2]
```diff
@@ -18,42 +19,45 @@ var Controller = function(){
...
 
-  //create dummy campaigns when db is empty ->
-  defaultCtrl.isEmpty( function(yes){
-    if( yes === true ){
-      var Template = {
-          name: 'Campaign-',
-          cre: { user: 'tmt'},
-          tcs: JSON.stringify({})
-        }
+  if( nconf.get('seeds') ){
```
Comment (jupe) [inline app/controllers/campaigns.js]: remove whole dummy data section because its not in use

File: app/controllers/resources.js [hunk 3]
```diff
@@ -27,45 +28,48 @@ var Controller = function(){
...
-          type: 'dut',
-          usage: {type: 'automation'},
-          status: {  value: 'active' },
-          other_info: { location: { site: 'Oulu'}, group: 'department' },
-          device: { 
-            manufacturer: 'Atmel',
-            model: 'sam4eXplained',
-            sn: '123568',
-            build: '5213'
+  if( nconf.get('seeds') ){
```
Comment (jupe) [inline app/controllers/resources.js]: remove whole dummy data section because its not in use

File: app/controllers/results.js [hunk 4]
```diff
@@ -24,7 +25,7 @@ var Controller = function(){
   var Testcase = mongoose.model('Testcase')
   var defaultCtrl = new DefaultController(Result, 'Result');
 
-  var doDummy = function() {
+  if( nconf.get('seeds') ){
```
Comment (jupe) [inline app/controllers/results.js]: remove whole dummy data section because its not in use

File: app/controllers/results.js [hunk 5]
```diff
@@ -162,10 +163,8 @@ var Controller = function(){
     });
 
   };
-  
-  //util.inherits(this, defaultCtrl);
 
-  doDummy();
+  //util.inherits(this, defaultCtrl);
```
Comment (jupe) [inline app/controllers/results.js]: remove commented line

File: app/controllers/results.js [hunk 6]
```diff
@@ -135,7 +136,7 @@ var Controller = function(){
             res.json({ok: 1, message: "created "+results.tests.length + " results"} );
           }
       });
-      
+
```
Comment (jupe) [inline app/controllers/results.js]: extra line feed

File: app/controllers/targets.js [hunk 7]
```diff
@@ -19,47 +19,49 @@ var Controller = function(){
...
-          binary: {
-            type: ".bin",
-          },
-          flash: {
-            method: "cp",
-            cycle_s: 4
-          },
-          reset: {
-            method: "default"
+  if( nconf.get('seeds') ){
```
Comment (jupe) [inline app/controllers/targets.js]: remove whole dummy data section because its not in use

File: app/controllers/testcases.js [hunk 8]
```diff
@@ -19,52 +20,54 @@ var Controller = function(){
   var Testcase = mongoose.model('Testcase');
   var defaultCtrl = new DefaultController(Testcase, 'testcase');
 
+  if( nconf.get('seeds') ){
```
Comment (jupe) [inline app/controllers/testcases.js]: remove whole dummy data section because its not in use

File: app/controllers/testcases.js [hunk 9]
```diff
@@ -19,52 +20,54 @@ var Controller = function(){
...
   this.paramFormat = defaultCtrl.format();
   this.paramTestcase = defaultCtrl.modelParam();
 
   this.all = function(req, res, next){
     // dummy middleman function..
-    next(); 
+    next();
   }
-  
+
```
Comment (jupe) [inline app/controllers/testcases.js]: extra line feed

File: config/express.js [hunk 10]
```diff
@@ -33,53 +33,44 @@ var env = process.env.NODE_ENV || 'development';
...
         skip = true;
         if( req.url.match(/^\/api/)!==null ) skip = false;
         return skip;
-      } 
+      }
     }));
-  
-  
-  
+
```
Comment (jupe) [inline config/express.js]: extra line feeds

File: Gruntfile.js [hunk 11]
```diff
@@ -14,7 +14,7 @@ module.exports = function(grunt) {
                     script: "index.js",
                     timeout: 3000,
                     harmony: true,
-                    args: ['-s'] //to more traces set -vvv instead of -s (silent)
+                    args: ['-vvv'] //to more traces set -vvv instead of -s (silent)
```
Comment (jupe) [inline Gruntfile.js]: I would not like to get anything to console except test framework prints -> if this still affect to that please revert back to `-s` .

File: app/controllers/apikeys.js [hunk 12]
```diff
@@ -1,41 +1,44 @@
+var _ = require('lodash');
```
Comment (jupe) [inline app/controllers/apikeys.js]: const

File: app/controllers/builds.js [hunk 13]
```diff
@@ -2,79 +2,38 @@
...
 var mongoose = require('mongoose');
-var _ = require('lodash');
 
-//own modules
+// own modules
 var DefaultController = require('./');
 
-var Controller = function(){
-
+var Controller = function () {
```
Comment (jupe) [inline app/controllers/builds.js]: let

File: app/controllers/index.js [hunk 14]
```diff
@@ -1,145 +1,142 @@
...
-    item.save( function(error){
-      if(error) {
+    item.save((error) => {
+      if (error) {
         winston.warn(error);
-        if(res) res.status(300).json({error: error});
+        if (res) res.status(300).json({ error: error });
+      } else if (res) {
+        req.query = req.body;
+        self.emit('create', item.toObject());
```
Comment (jupe) [inline app/controllers/index.js]: this.emit

File: app/controllers/index.js [hunk 15]
```diff
@@ -1,145 +1,142 @@
...
+    Model.findByIdAndUpdate(req.params[defaultModelName], req.body, (error, doc) => {
+      if (error) {
+        res.status(300).json({ error: error });
       } else {
-        if(res){
-          req.query = req.body;
-          self.emit('create', item.toObject());
-          res.json(item);
-        }
+        self.emit('update', doc.toObject());
```
Comment (jupe) [inline app/controllers/index.js]: this.emit

File: app/controllers/index.js [hunk 16]
```diff
@@ -1,145 +1,142 @@
...
-          self.emit('remove', req.params[defaultModelName]);
-          res.json({});
-        } 
-      });
-  }
+    Model.findByIdAndRemove(find, (error, ok) => {
+      if (error) {
+        res.status(300).json({ error: error });
+      } else {
+        self.emit('remove', req.params[defaultModelName]);
```
Comment (jupe) [inline app/controllers/index.js]: this.emit

File: app/controllers/index.js [hunk 17]
```diff
@@ -1,145 +1,142 @@
...
-    if( count > 0 ){
-      var o = new Model( doItem(count) );
-      o.save( function(err){
-        if( err ) done(err );
-        else self.generateDummyData( doItem, count-1, done );
+  };
+
+  this.generateDummyData = (doItem, count, done) => {
+    if (count > 0) {
+      var o = new Model(doItem(count));
```
Comment (jupe) [inline app/controllers/index.js]: let

File: app/controllers/items.js [hunk 18]
```diff
@@ -2,112 +2,114 @@
...
-  // Regular save
-  req.Item.save(function(err) {
-	if (err) { 
-	  winston.error(err.message);
-	  return res.status(400).json({error:err.message}); 
-	}
-	else {
-	  res.status(200).json(req.Loan); 
-	}
+    return undefined;
```
Comment (jupe) [inline app/controllers/items.js]: is this really needed ?

File: app/controllers/items.js [hunk 19]
```diff
@@ -2,112 +2,114 @@
...
   });
 }
 
 function handleUpdateInStock(req, res) {
-  var delta_stock = req.body.in_stock - req.Item.in_stock;
+  var deltaStock = req.body.in_stock - req.Item.in_stock;
```
Comment (jupe) [inline app/controllers/items.js]: let

File: app/controllers/items.js [hunk 20]
```diff
@@ -2,112 +2,114 @@
...
   // Increase availability with the same amount that in_stock was changed
-  winston.info('Received only in_stock in PUT, automatically modifying available with the same amount:' + delta_stock);
-  req.body.available = req.Item.available + delta_stock;
+  winston.info('Received only in_stock in PUT, automatically modifying available with the same amount:' + deltaStock);
+  req.body.available = req.Item.available + deltaStock;
 }
 
 function handleUpdateAvailable(req, res) {
-  var delta_stock = req.body.available - req.Item.available;
+  var deltaStock = req.body.available - req.Item.available;
```
Comment (jupe) [inline app/controllers/items.js]: let

File: app/controllers/loans.js [hunk 21]
```diff
@@ -2,119 +2,125 @@
...
-	  winston.info('Update completed successfully');
-	  res.status(200).json(req.Loan);
-	}
-  });  
-}
 function handleItemsInUpdate(req, res) {
-  var count_array = req.Loan.countReturns(req.body.items)
-  if (count_array instanceof Error) {
-    return count_array;
+  var countArray = req.Loan.countReturns(req.body.items);
```
Comment (jupe) [inline app/controllers/loans.js]: let

File: app/controllers/loans.js [hunk 22]
```diff
@@ -2,119 +2,125 @@
...
+    return undefined;
+  });
+
+  return undefined;
+}
+
+function customUpdate(req, res) {
+  if (req.body.items !== undefined) {
+    winston.info('PUT request with items property received');
+    var err = handleItemsInUpdate(req, res);
```
Comment (jupe) [inline app/controllers/loans.js]: let

File: app/controllers/loans.js [hunk 23]
```diff
@@ -2,119 +2,125 @@
...
+      winston.error(err.message);
+      return res.status(400).json({ error: err.message });
+    }
+
+    return res.status(200).json({});
   });
 }
 
+var Controller = function () {
+  var defaultCtrl = new DefaultController(Loan, 'Loan');
```
Comment (jupe) [inline app/controllers/loans.js]: let

File: app/models/item.js [hunk 24]
```diff
@@ -7,7 +7,7 @@ var ObjectId = Types.ObjectId;
 var QueryPlugin = require('mongoose-query');
 
 var ItemSchema = new Schema({
-  barcode : { type: String, unique:true, sparse:true },
+  barcode : { type: String, required:true, unique:true },
```
Comment (jupe) [inline app/models/item.js]: this cannot be required I think

File: app/controllers/items.js (hunk 18 above)
Comment (olkorhon) [inline app/controllers/items.js]: Serves no functional purpose, but EsLint enforces a rule which states that if a function returns anything, it must return something in all cases.

File: app/models/item.js (hunk 24 above)
Comment (jupe) [inline app/models/item.js]: revert this back to original

Comment (jupe) [inline app/models/item.js]: change only this one for now. rest of changes in another PR

**Pull request merged**